src/nifs/protogalaxy/mod.rs

```diff
@@ -139,3 +149,135 @@ impl<C: CurveAffine> MultifoldingScheme<C> for ProtoGalaxy<C> {
+/// compute weighted sum: sum_i L_i(gamma)w_i
+/// Here w_i is witness, vector of vector
+/// # Parameters
+///
+/// - `gamma` - eval Lagrange polynomials at this point
+/// - `log_n` - `log2(n+1)`, where `n` - number of instances to be folded
+/// - `ws` - `{w_0,..., w_n}`, w_i is vector of vector of F
+///
+/// # Result
+///
+/// the vector of weighted sum
+/// `L_0(gamma)w_0 + L_1(gamma)w_1, ..., L_n(gamma)w_n`
+fn fold_witnesses<F: PrimeField>(gamma: F, log_n: usize, wss: &[&[Vec<F>]]) -> Vec<Vec<F>> {
+    let ll: Vec<F> = iter_eval_lagrange_polynomials_for_cyclic_group(gamma, log_n).collect();
+    (0..wss[0].len())
+        .into_par_iter()
+        .map(|ii| {
+            (0..wss[0][ii].len())
+                .into_par_iter()
+                .map(|jj| {
+                    ll.iter()
+                        .zip(wss.iter())
+                        .map(|(l_k, w_k)| *l_k * w_k[ii][jj])
+                        .fold(F::ZERO, |sum, val| sum + val)
+                })
+                .collect()
+        })
+        .collect()
+}
+
+fn evaluate_G_poly<F: PrimeField>(
+    beta: F,
+    gamma: F,
+    S: &PlonkStructure<F>,
+    traces: &[(impl GetChallenges<F> + GetWitness<F>)],
+) -> Result<F, eval::Error> {
+    let count_of_rows = 1 << S.k;
+    let count_of_gates = S.gates.len();
+    let n = count_of_rows * count_of_gates;
+    let t = n.next_power_of_two().ilog2() as usize;
+
+    let num_instances = traces.len().next_power_of_two();
+    let witnesses: Vec<&[Vec<F>]> = traces.iter().map(|trace| trace.get_witness()).collect();
+    let folded_w = fold_witnesses(gamma, num_instances, &witnesses);
+    let challenges: Vec<&[F]> = traces.iter().map(|trace| trace.get_challenges()).collect();
+    let folded_c = fold_challenges(gamma, num_instances, &challenges);
+
+    let evaluated = S.gates.iter().flat_map(|gate| {
+        let eval_domain = PlonkEvalDomain {
+            num_advice: S.num_advice_columns,
+            num_lookup: S.num_lookups(),
+            selectors: &S.selectors,
+            fixed: &S.fixed_columns,
+            challenges: &folded_c,
+            W1s: &folded_w,
+            W2s: &[],
+        };
+
+        let evaluator = GraphEvaluator::new(gate);
+        (0..eval_domain.row_size())
+            .map(move |row_index| evaluator.evaluate(&eval_domain, row_index))
+    });
+    iter_eval_of_pow_i(
+        NonZeroUsize::new(t).unwrap(),
+        NonZeroUsize::new(n).unwrap(),
+        beta,
+    )
+    .zip(evaluated)
+    .map(|(beta_i, f_i)| {
+        let f_i = f_i?;
+        Ok(beta_i * f_i)
+    })
+    .try_fold(F::ZERO, |sum, part| Ok(sum + part?))
+}
+
+fn compute_G_poly<F: PrimeField>(
+    // number of instances to be folded
+    k: usize,
+    // beta: the folded beta value
+    beta: F,
+    S: &PlonkStructure<F>,
+    // first one is accumulator, the rest k traces are instances to be folded
+    traces: &[(impl GetChallenges<F> + GetWitness<F>)],
```

Comment:
The acc type will be different and this impl will not work, and it will be clearer if it is a separate argument

Reply:
the acc type doesn't have the Get traits implemented? we can add them so the treatment of them are uniform

Reply:
When you write `&[impl ...]`, impl outputs one type and the whole slice must have that type.

This is static, not dynamic at heap, so if you have the first element of one type and the rest of the slice of another type, the compiler will not agree with it